### PR TITLE
hw8.1

### PR DESCRIPTION
**Pull request by Mykhaylo12**

### CODE REVIEW

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 1]
```diff
@@ -3,48 +3,161 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int amountOfElements = 0;
```
[the remark this answers is not in the document]
Comment (Mykhaylo12) [reply src/main/java/core/basesyntax/MyLinkedList.java]: ok

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 2]
```diff
@@ -3,48 +3,161 @@
...
+    private Box<T> firstBox;
```
[the remark this answers is not in the document]
Comment (Mykhaylo12) [reply src/main/java/core/basesyntax/MyLinkedList.java]: ok

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 3]
```diff
@@ -3,48 +3,161 @@
...
+    private Box<T> lastBox;
```
[the remark this answers is not in the document]
Comment (Mykhaylo12) [reply src/main/java/core/basesyntax/MyLinkedList.java]: ok

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 4]
```diff
@@ -3,48 +3,161 @@
...
+                iteratorBox = iteratorBox.linkToPrevBox;
+                count--;
+            }
+            return iteratorBox;
+        }
+    }
+
+    private static class Box<T> {
+        T value;
+        Box<T> linkToNextBox;
```
[the remark this answers is not in the document]
Comment (Mykhaylo12) [reply src/main/java/core/basesyntax/MyLinkedList.java]: ok

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 5]
```diff
@@ -3,48 +3,161 @@
...
+            while (count > index) {
+                iteratorBox = iteratorBox.linkToPrevBox;
+                count--;
+            }
+            return iteratorBox;
+        }
+    }
+
+    private static class Box<T> {
+        T value;
```
[the remark this answers is not in the document]
Comment (Mykhaylo12) [reply src/main/java/core/basesyntax/MyLinkedList.java]: ok

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 6]
```diff
@@ -3,48 +3,161 @@
...
+                firstBox = newBox;
+                amountOfElements++;
+            }
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
```
[the remark this answers is not in the document]
Comment (Mykhaylo12) [reply src/main/java/core/basesyntax/MyLinkedList.java]: thank you

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 7]
```diff
@@ -3,48 +3,161 @@
...
+            int count = amountOfElements - 1;
+            while (count > index) {
+                iteratorBox = iteratorBox.linkToPrevBox;
+                count--;
+            }
+            return iteratorBox;
+        }
+    }
+
+    private static class Box<T> {
```
[the remark this answers is not in the document]
Comment (Mykhaylo12) [reply src/main/java/core/basesyntax/MyLinkedList.java]: do I need to change or it is only suggestion?

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 8]
```diff
@@ -3,48 +3,161 @@
...
+    private Box<T> firstBox;
+    private Box<T> lastBox;
+
     @Override
     public void add(T value) {
-
+        if (amountOfElements == 0) {
+            Box newBox = new Box(value, null, null);
+            firstBox = lastBox = newBox;
+            amountOfElements++;
```
[the remark this answers is not in the document]
Comment (Mykhaylo12) [reply src/main/java/core/basesyntax/MyLinkedList.java]: ok

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 9]
```diff
@@ -3,48 +3,161 @@
...
+                count--;
+            }
+            return iteratorBox;
+        }
+    }
+
+    private static class Box<T> {
+        T value;
+        Box<T> linkToNextBox;
+        Box<T> linkToPrevBox;
```
[the remark this answers is not in the document]
Comment (Mykhaylo12) [reply src/main/java/core/basesyntax/MyLinkedList.java]: ok

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (Mykhaylo12) [reply src/main/java/core/basesyntax/MyLinkedList.java]: for  uderstanding better to keep it Box

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 10]
```diff
@@ -3,48 +3,161 @@
...
+        if (index < 0 || index > amountOfElements) {
+            throw new IndexOutOfBoundsException();
+        }
+        if (index == amountOfElements) {
+            add(value);
+        } else {
+            Box<T> boxToMoveAhead = searchByIndex(index);
+            Box newBox = new Box(value, boxToMoveAhead, boxToMoveAhead.linkToPrevBox);
+            boxToMoveAhead.linkToPrevBox = newBox;
+            newBox.linkToNextBox.linkToNextBox = newBox;
```
[the remark this answers is not in the document]
Comment (Mykhaylo12) [reply src/main/java/core/basesyntax/MyLinkedList.java]: ok

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 11]
```diff
@@ -105,59 +104,59 @@ public T remove(T t) {
...
     private static class Box<T> {
-        T value;
-        Box<T> linkToNextBox;
-        Box<T> linkToPrevBox;
+        private T value;
+        private Box<T> next;
+        private Box<T> previous;
 
-        Box(T value, Box<T> linkToNextBox, Box<T> linkToPrevBox) {
+        Box(T value, Box<T> next, Box<T> previous) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: модифікатор

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (Sofasmile) [reply src/main/java/core/basesyntax/MyLinkedList.java]: ок

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 12]
```diff
@@ -3,48 +3,160 @@
...
-
+        if (size == 0) {
+            Box newBox = new Box(value, null, null);
+            head = last = newBox;
+            size++;
+        } else {
+            Box newBox = new Box(value, null, last);
+            last.next = newBox;
+            last = newBox;
+            size++;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: ми можемо винести size++; після   if-else

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 13]
```diff
@@ -3,48 +3,160 @@
...
+            if (interatorBox.value == value || value != null & value.equals(interatorBox.value)) {
+                return count;
+            }
+            interatorBox = interatorBox.next;
+            count++;
+        }
+        return size;
+    }
+
+    public Box searchByIndex(int index) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]:
```suggestion
    public Box <T> searchByIndex(int index) {
```

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 14]
```diff
@@ -3,48 +3,160 @@
...
+        }
+        Box<T> boxToChangeValue = searchByIndex(index);
+        boxToChangeValue.value = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Box<T> boxToRemove = searchByIndex(index);
+        final Object elementToRemove = boxToRemove.value;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: Object???

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 15]
```diff
@@ -3,48 +3,160 @@
...
+        } else if (boxToRemove == last) {
+            last = boxToRemove.previous;
+            last.next = null;
+            size--;
+            return (T) elementToRemove;
+        } else {
+            boxToRemove.previous.next = boxToRemove.next;
+            boxToRemove.next.previous = boxToRemove.previous;
+            size--;
+            return (T) elementToRemove;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]:
```suggestion
            return elementToRemove;
```

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 16]
```diff
@@ -3,48 +3,160 @@
...
-        return null;
+        Box<T> boxToRemove = searchByIndex(index);
+        final Object elementToRemove = boxToRemove.value;
+        if (index < 0 || size < index) {
+            throw new IndexOutOfBoundsException();
+        }
+        if (size == 1) {
+            head = last = null;
+            size--;
+            return (T) elementToRemove;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]:
```suggestion
            return elementToRemove;
```
the same everywhere

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 11 above)
Comment (Mykhaylo12) [reply src/main/java/core/basesyntax/MyLinkedList.java]: ok

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 12 above)
Comment (Mykhaylo12) [reply src/main/java/core/basesyntax/MyLinkedList.java]: done

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 14 above)
Comment (Mykhaylo12) [reply src/main/java/core/basesyntax/MyLinkedList.java]: T

**Pull request closed without merging**